gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java

```diff
@@ -121,28 +122,34 @@ public TaskResult run() {
         case CANCELLED:
           return new TaskResult(TaskResult.Status.CANCELED, "task id: " + this.taskId);
         default:
-          throw new IllegalStateException("Unexpected result WorkUnit state: " + workingState);
+          throw new IllegalStateException("Unexpected result task state: " + workingState);
       }
     } catch (IOException ioe) {
-      LOGGER.error("Failed to deserialize WorkUnit " + this.taskId, ioe);
-      throw Throwables.propagate(ioe);
+      LOGGER.error("Failed to build task " + this.taskId, ioe);
+      return new TaskResult(TaskResult.Status.ERROR, "task id: " + this.taskId);
     } catch (ExecutionException ee) {
-      LOGGER.error("Failed to run WorkUnit " + this.taskId, ee);
-      throw Throwables.propagate(ee);
+      LOGGER.error("Failed to run task " + this.taskId, ee);
+      return new TaskResult(TaskResult.Status.ERROR, "task id: " + this.taskId);
     } catch (InterruptedException ie) {
-      LOGGER.error("Failed to run WorkUnit " + this.taskId, ie);
-      throw Throwables.propagate(ie);
+      cancel();
+      Thread.currentThread().interrupt();
+      return new TaskResult(TaskResult.Status.CANCELED, "task id: " + this.taskId);
+    } catch (CancellationException ce) {
+      return new TaskResult(TaskResult.Status.CANCELED, "task id: " + this.taskId);
+    } catch (Throwable t) {
+      LOGGER.error("Failed to run task " + this.taskId, t);
+      return new TaskResult(TaskResult.Status.ERROR, "task id: " + this.taskId);
     }
   }
 
   @Override
   public void cancel() {
-    LOGGER.info(String.format("Cancelling WorkUnit %s of job %s", this.taskId, this.jobId));
+    LOGGER.info(String.format("Cancelling task %s of job %s", this.taskId, this.jobId));
     this.futureOptional.transform(new Function<Future<?>, Object>() {
       @Nullable
       @Override
       public Object apply(Future<?> input) {
-        return input.cancel(true);
+        return input.cancel(false);
```

Comment:
If an `InterruptedException` is caught in line 133, shouldn't the task be cancelled asap? Also what about hanging tasks?